### PR TITLE
Explicit raise on invalid model_index + add `ignore_metadata_errors` option

### PR DESCRIPTION
**Pull request by Wauplin**
Fix https://github.com/huggingface/huggingface_hub/issues/1361 cc @davanstrien 

As suggested by @davanstrien we should raise an explicit issue + add an option if a user _really_ wants to ignore parsing error (=> better to fail by default than losing some information). For now option is used purely in `ModelCard` but also implemented in `DatasetCard` and `SpaceCard` as we're using inheritance.

This is a breaking change but as discussed on [slack](https://huggingface.slack.com/archives/C02V5EA0A95/p1677159836698829) (internal link), model cards with invalid metadata are already rejected by the server when pushed on the Hub. So this change should only affect invalid metadata on legacy model cards on the Hub.

Adapted the tests accordingly.
(cc @lvwerra who since I know you're using `model_index`)

**EDIT:** also refactored some tests to get rid of `Repository` (being quite slower than HfApi)

### CODE REVIEW

**Comment by davanstrien** [comment]
These changes look good. My main question is whether it makes more sense to set `ignore_metadata_errors=True` for some time period to delay the breaking change:

> This is a breaking change but as discussed on [slack](https://huggingface.slack.com/archives/C02V5EA0A95/p1677159836698829) (internal link), model cards with invalid metadata are already rejected by the server when pushed on the Hub. So this change should only affect invalid metadata on legacy model cards on the Hub. 

This makes sense to me; my only question is we know of other libraries using `ModelCard` for creating model metadata? If there are it might be worth trying to upstream fixes to `moded-index` before including this in a release? IIRC `SpaCy` wrote the `model-index` directly so wouldn't be impacted.

**Comment by Wauplin** [comment]
> my only question is if we know of other libraries using ModelCard for creating model metadata?

I'm not aware of a lot of integrations using it yet but doesn't mean there aren't any. [skops](https://github.com/skops-dev/skops/blob/5b3a7a4f0d8bc00030f96f0c53391a5baf047e62/skops/card/_model_card.py#L468) is using it but it's not generating `model-index` itself. As you said, [spacy integration](https://github.com/explosion/spacy-huggingface-hub/blob/c76ac641e24789ff344b73b0eb5dfa6d67fa6a51/spacy_huggingface_hub/push.py#L222) was problematic but didn't use huggingface_hub to generate them. I know `evaluate` uses them (@lvwerra) but there shouldn't be any issue as it's the main package we have to generate results (i.e. we would already know about it). Maybe @osanseviero you can think about another lib?

Otherwise I'm still convinced it's better to consider that as "fixing a bug" rather than "introducing a breaking change", even though both are factually true.

**Comment by Wauplin** [comment]
BTW, PR is ready for review. Failing tests are unrelated.

File: src/huggingface_hub/repocard_data.py [hunk 1]
```diff
@@ -421,6 +436,9 @@ class SpaceCardData(CardData):
             List of datasets related to this Space. Should be a dataset ID found on https://hf.co/datasets.
         tags (`List[str]`, *optional*)
             List of tags to add to your Space that can be used when filtering on the Hub.
+        ignore_metadata_errors (`str`):
+            If True, errors while parsing the metadata section will be ignored. Some information might be lost during
+            the process. Use it at your own risks.
```
Comment (davanstrien) [inline src/huggingface_hub/repocard_data.py:441]:
```suggestion
            the process. Use it at your own risk.
```

**Review by davanstrien — COMMENTED** [review]
added a few nit suggestions but otherwise looks good to me.

**Review by LysandreJik — APPROVED** [review]
Looks good to me, coherent change

**Comment by Wauplin** [comment]
Thanks for the reviews @davanstrien @nateraw @LysandreJik ! I'm merging it :)

**Pull request merged**